File: server/src/main/java/dev/slimevr/osc/VRMReader.java
package dev.slimevr.osc;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.jme3.math.Vector3f;
import dev.slimevr.tracking.processor.BoneType;
import io.eiren.util.logging.LogManager;

import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Map;


public class VRMReader {
	private final String vrmPath;
	private final ObjectMapper objectMapper;

	public VRMReader(String vrmPath) {
//		vrmPath = "C:/Users/mario/Documents/VSeeFace Custom Data/Auy2.3.1.vrm";
		this.vrmPath = vrmPath;
		objectMapper = new ObjectMapper(new JsonFactory());
		objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
	}

	public Vector3f getOffsetForBone(BoneType boneType) {
		Vector3f translation = new Vector3f();

		// extensions.VRM.humanoid.humanBones
		// "bone" = UnityBone
		// read "node"

		// nodes
		// get "translation" value at "node" place of the table
		// Inverse z axis

		try {
			// convert JSON file to map
			Map<?, ?> map = (Map<?, ?>) Arrays
				.asList(objectMapper.readValue(Paths.get(vrmPath).toFile(), Map[].class));

			// print map entries
			for (Map.Entry<?, ?> entry : map.entrySet()) {
				System.out.println(entry.getKey() + "=" + entry.getValue());
			}

		} catch (Exception ex) {
			LogManager.warning("[VMCReader] Error parsing \"" + vrmPath + "\"");
		}


		return translation;
	}
}

File: server/src/main/java/dev/slimevr/osc/VRMReader.kt
package dev.slimevr.osc

import com.jme3.math.Vector3f
import dev.slimevr.tracking.trackers.UnityBone
import io.eiren.util.logging.LogManager
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import java.util.*

private val jsonIgnoreKeys = Json { ignoreUnknownKeys = true }
class VRMReader(vrmJson: String) {
	private val data: GLTF = jsonIgnoreKeys.decodeFromString(vrmJson)

	fun getOffsetForBone(unityBone: UnityBone): Vector3f {
		val translation = Vector3f()

		val bone = try {
			data.extensions.vrm.humanoid.humanBones.first { it.bone.equals(unityBone.stringVal, ignoreCase = true) }
		} catch (e: NoSuchElementException) {
			LogManager.warning("[VRMReader] Bone ${unityBone.stringVal} not found in JSON")
			return translation
		}

		val translationNode = data.nodes[bone.node].translation ?: return translation
		translation.x = translationNode[0].toFloat()
		translation.y = translationNode[1].toFloat()
		translation.z = -translationNode[2].toFloat()
		if (unityBone != UnityBone.LEFT_FOOT && unityBone != UnityBone.RIGHT_FOOT) {
			translation.z = -translation.z
		}

		return translation
	}
}

@Serializable
data class GLTF(
	val extensions: Extensions,
	val extensionsUsed: List<String>,
	val nodes: List<Node>,
)

@Serializable
data class Extensions(
	@SerialName("VRM")
	val vrm: VRM,
)

@Serializable
data class VRM(
	val humanoid: Humanoid,
)

@Serializable
data class Humanoid(
	val humanBones: List<HumanBone>,
	val armStretch: Double,
	val legStretch: Double,
	val upperArmTwist: Double,
	val lowerArmTwist: Double,
	val upperLegTwist: Double,
	val lowerLegTwist: Double,
	val feetSpacing: Double,
	val hasTranslationDoF: Boolean,
)

@Serializable
data class HumanBone(
	val bone: String,
	val node: Int,
	val useDefaultValues: Boolean,
)

@Serializable
data class Node(
	val translation: List<Double>? = null,
	val rotation: List<Double>? = null,
	val scale: List<Double>? = null,
	// GLTF says that there can be a matrix instead of translation,
	// rotation and scale, so we need to support that too in the future.
	// val matrix: List<List<Double>>,
	val children: List<Int> = emptyList(),
)
